### ISSUE TITLE
Crash when FileCapture closes without running to completion.

### ISSUE THREAD

**Issue by markgras** [KimiNewt/pyshark#548]
This is a repeat of issue #395, which was closed due to old age but was never fixed.

## Environment info
- Python 3.10.5 64-Bit
- Pyshark 0.4.6
- Windows and Tshark info:
```
C:\WINDOWS\system32>"C:/Program Files/Wireshark/tshark.exe" --version
TShark (Wireshark) 3.6.5 (v3.6.5-0-g21f79ddbefbd)

Copyright 1998-2022 Gerald Combs <gerald@wireshark.org> and contributors.
License GPLv2+: GNU GPL version 2 or later <https://www.gnu.org/licenses/gpl-2.0.html>
This is free software; see the source for copying conditions. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Compiled (64-bit) using Microsoft Visual Studio 2019 (VC++ 14.31, build 31107),
with libpcap, with GLib 2.66.4, with zlib 1.2.11, with Lua 5.2.4, with GnuTLS
3.6.3 and PKCS #11 support, with Gcrypt 1.8.3, with MIT Kerberos, with MaxMind
DB resolver, with nghttp2 1.44.0, with brotli, with LZ4, with Zstandard, with
Snappy, with libxml2 2.9.10, with libsmi 0.4.8.

Running on 64-bit Windows 10 (21H2), build 19044, with Intel(R) Core(TM)
i7-10875H CPU @ 2.30GHz (with SSE4.2), with 32512 MB of physical memory, with
GLib 2.66.4, with Npcap version 1.55, based on libpcap version 1.10.2-PRE-GIT,
with c-ares 1.17.0, with GnuTLS 3.6.3, with Gcrypt 1.8.3, with nghttp2 1.44.0,
with brotli 1.0.9, with LZ4 1.9.3, with Zstandard 1.4.0, with
LC_TYPE=English_United States.utf8, binary plugins supported (0 loaded).
```

## Issue
Tshark crashes when the pyshark file capture closes without running to completion. Two flavors of code are shown below.

Flavor 1:
```
import pyshark


def iterate_through_pcap(file_name):
    cap = pyshark.FileCapture(file_name, keep_packets=False)
    cap.set_debug()
    try:
        for i, pkt in enumerate(cap):

            # Parse packet
            pass

            # Yield result
            yield i
    finally:
        cap.close()


if __name__ == '__main__':
    result = iterate_through_pcap('test0.pcapng')
    print(next(result))
```

Traceback 1:
```
C:\Users\mgrassi\Desktop\venvs\pyshark\Scripts\python.exe C:/Users/mgrassi/Desktop/test/pyshark_error.py
2022-06-07 09:56:57,389 - FileCapture - DEBUG - Creating TShark subprocess with parameters: C:/Program Files/Wireshark/tshark.exe -l -n -T pdml -r test0.pcapng
2022-06-07 09:56:57,389 - FileCapture - DEBUG - Executable: C:/Program Files/Wireshark/tshark.exe
2022-06-07 09:56:57,397 - FileCapture - DEBUG - TShark subprocess (pid 7560) created
0
2022-06-07 09:56:57,847 - FileCapture - DEBUG - Cleanup Subprocess (pid 7560)
Exception in callback _ProactorReadPipeTransport._loop_reading(<_OverlappedF... result=12191>)
handle: <Handle _ProactorReadPipeTransport._loop_reading(<_OverlappedF... result=12191>)>
Traceback (most recent call last):
  File "C:\Users\mgrassi\Desktop\venvs\pyshark\lib\site-packages\pyshark\capture\capture.py", line 268, in _packets_from_tshark_sync
    yield packet
GeneratorExit

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 301, in _loop_reading
    self._read_fut = self._loop._proactor.recv_into(self._sock, self._data)
  File "C:\Program Files\Python310\lib\asyncio\windows_events.py", line 473, in recv_into
    self._register_with_iocp(conn)
  File "C:\Program Files\Python310\lib\asyncio\windows_events.py", line 723, in _register_with_iocp
    _overlapped.CreateIoCompletionPort(obj.fileno(), self._iocp, 0, 0)
OSError: [WinError 87] The parameter is incorrect

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\asyncio\events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 311, in _loop_reading
    self._fatal_error(exc, 'Fatal read error on pipe transport')
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 131, in _fatal_error
    self._force_close(exc)
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 134, in _force_close
    if self._empty_waiter is not None and not self._empty_waiter.done():
AttributeError: '_ProactorReadPipeTransport' object has no attribute '_empty_waiter'
Error on reading from the event loop self pipe
loop: <ProactorEventLoop running=True closed=False debug=False>
Traceback (most recent call last):
  File "C:\Users\mgrassi\Desktop\venvs\pyshark\lib\site-packages\pyshark\capture\capture.py", line 268, in _packets_from_tshark_sync
    yield packet
GeneratorExit

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 780, in _loop_self_reading
    f = self._proactor.recv(self._ssock, 4096)
  File "C:\Program Files\Python310\lib\asyncio\windows_events.py", line 450, in recv
    self._register_with_iocp(conn)
  File "C:\Program Files\Python310\lib\asyncio\windows_events.py", line 723, in _register_with_iocp
    _overlapped.CreateIoCompletionPort(obj.fileno(), self._iocp, 0, 0)
OSError: [WinError 87] The parameter is incorrect
Task was destroyed but it is pending!
task: <Task pending name='Task-9' coro=<Capture.close_async() running at C:\Users\mgrassi\Desktop\venvs\pyshark\lib\site-packages\pyshark\capture\capture.py:463>>
C:\Program Files\Python310\lib\asyncio\base_events.py:671: RuntimeWarning: coroutine 'Capture.close_async' was never awaited
Task was destroyed but it is pending!
task: <Task pending name='Task-10' coro=<Capture.close_async() running at C:\Users\mgrassi\Desktop\venvs\pyshark\lib\site-packages\pyshark\capture\capture.py:463>>
Exception ignored in: <function BaseSubprocessTransport.__del__ at 0x000001AA3E167EB0>
Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\asyncio\base_subprocess.py", line 126, in __del__
  File "C:\Program Files\Python310\lib\asyncio\base_subprocess.py", line 104, in close
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 108, in close
  File "C:\Program Files\Python310\lib\asyncio\base_events.py", line 750, in call_soon
  File "C:\Program Files\Python310\lib\asyncio\base_events.py", line 515, in _check_closed
RuntimeError: Event loop is closed
Exception ignored in: <function _ProactorBasePipeTransport.__del__ at 0x000001AA3E18D900>
Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 115, in __del__
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 79, in __repr__
  File "C:\Program Files\Python310\lib\asyncio\windows_utils.py", line 102, in fileno
ValueError: I/O operation on closed pipe

Process finished with exit code 0
```

Flavor 2:
```
import pyshark


def iterate_through_pcap(file_name):
    with pyshark.FileCapture(file_name, keep_packets=False) as cap:
        cap.set_debug()
        for i, pkt in enumerate(cap):

            # Parse packet
            pass

            # Yield result
            yield i


if __name__ == '__main__':
    result = iterate_through_pcap('test0.pcapng')
    print(next(result))
```

Traceback 2:
```
C:\Users\mgrassi\Desktop\venvs\pyshark\Scripts\python.exe C:/Users/mgrassi/Desktop/test/pyshark_error.py
2022-06-07 09:54:45,213 - FileCapture - DEBUG - Creating TShark subprocess with parameters: C:/Program Files/Wireshark/tshark.exe -l -n -T pdml -r test0.pcapng
2022-06-07 09:54:45,213 - FileCapture - DEBUG - Executable: C:/Program Files/Wireshark/tshark.exe
2022-06-07 09:54:45,218 - FileCapture - DEBUG - TShark subprocess (pid 16500) created
0
2022-06-07 09:54:45,665 - FileCapture - DEBUG - Cleanup Subprocess (pid 16500)
Exception in callback _ProactorReadPipeTransport._loop_reading(<_OverlappedF... result=12191>)
handle: <Handle _ProactorReadPipeTransport._loop_reading(<_OverlappedF... result=12191>)>
Traceback (most recent call last):
  File "C:\Users\mgrassi\Desktop\venvs\pyshark\lib\site-packages\pyshark\capture\capture.py", line 268, in _packets_from_tshark_sync
    yield packet
GeneratorExit

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 301, in _loop_reading
    self._read_fut = self._loop._proactor.recv_into(self._sock, self._data)
  File "C:\Program Files\Python310\lib\asyncio\windows_events.py", line 473, in recv_into
    self._register_with_iocp(conn)
  File "C:\Program Files\Python310\lib\asyncio\windows_events.py", line 723, in _register_with_iocp
    _overlapped.CreateIoCompletionPort(obj.fileno(), self._iocp, 0, 0)
OSError: [WinError 87] The parameter is incorrect

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\asyncio\events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 311, in _loop_reading
    self._fatal_error(exc, 'Fatal read error on pipe transport')
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 131, in _fatal_error
    self._force_close(exc)
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 134, in _force_close
    if self._empty_waiter is not None and not self._empty_waiter.done():
AttributeError: '_ProactorReadPipeTransport' object has no attribute '_empty_waiter'
Error on reading from the event loop self pipe
loop: <ProactorEventLoop running=True closed=False debug=False>
Traceback (most recent call last):
  File "C:\Users\mgrassi\Desktop\venvs\pyshark\lib\site-packages\pyshark\capture\capture.py", line 268, in _packets_from_tshark_sync
    yield packet
GeneratorExit

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 780, in _loop_self_reading
    f = self._proactor.recv(self._ssock, 4096)
  File "C:\Program Files\Python310\lib\asyncio\windows_events.py", line 450, in recv
    self._register_with_iocp(conn)
  File "C:\Program Files\Python310\lib\asyncio\windows_events.py", line 723, in _register_with_iocp
    _overlapped.CreateIoCompletionPort(obj.fileno(), self._iocp, 0, 0)
OSError: [WinError 87] The parameter is incorrect
Task was destroyed but it is pending!
task: <Task pending name='Task-9' coro=<Capture.close_async() running at C:\Users\mgrassi\Desktop\venvs\pyshark\lib\site-packages\pyshark\capture\capture.py:463>>
C:\Program Files\Python310\lib\asyncio\base_events.py:671: RuntimeWarning: coroutine 'Capture.close_async' was never awaited
Task was destroyed but it is pending!
task: <Task pending name='Task-10' coro=<Capture.close_async() running at C:\Users\mgrassi\Desktop\venvs\pyshark\lib\site-packages\pyshark\capture\capture.py:463>>
Exception ignored in: <function BaseSubprocessTransport.__del__ at 0x00000282949C7EB0>
Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\asyncio\base_subprocess.py", line 126, in __del__
  File "C:\Program Files\Python310\lib\asyncio\base_subprocess.py", line 104, in close
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 108, in close
  File "C:\Program Files\Python310\lib\asyncio\base_events.py", line 750, in call_soon
  File "C:\Program Files\Python310\lib\asyncio\base_events.py", line 515, in _check_closed
RuntimeError: Event loop is closed
Exception ignored in: <function _ProactorBasePipeTransport.__del__ at 0x00000282949ED900>
Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 115, in __del__
  File "C:\Program Files\Python310\lib\asyncio\proactor_events.py", line 79, in __repr__
  File "C:\Program Files\Python310\lib\asyncio\windows_utils.py", line 102, in fileno
ValueError: I/O operation on closed pipe

Process finished with exit code 0
```

Both finish without error if `print(next(result))` is changed to `list(result)`, meaning that the file capture is completely parsed.

**Comment by KimiNewt** [comment]
Seems to only occur in Windows unsurprisingly. I don't have a machine to test it on handy but I'd first check if the completely async version of this code (using `close_async` , `packets_from_tshark` etc) still fails

**Comment by wlockwood** [comment]
I was able to resolve this by waiting for the tasks to finish with the following code:
```python
with pyshark.FileCapture(CAP_FILE, display_filter=SHARK_DISPLAY_FILTER) as cap_context:
    cap = list(cap_context)
pending = asyncio.all_tasks(cap_context.eventloop)
cap_context.eventloop.run_until_complete(asyncio.gather(*pending))
```

**Comment by wlockwood** [comment]
> Seems to only occur in Windows unsurprisingly. I don't have a machine to test it on handy but I'd first check if the completely async version of this code (using close_async , packets_from_tshark etc) still fails

I tried a number of things, such as awaiting `close_async` and calling `close` and then just waiting a few seconds, but it never seemed to work without the code in my previous comment.

**Comment by markgras** [comment]
> I was able to resolve this by waiting for the tasks to finish with the following code:
> 
> ```python
> with pyshark.FileCapture(CAP_FILE, display_filter=SHARK_DISPLAY_FILTER) as cap_context:
>     cap = list(cap_context)
> pending = asyncio.all_tasks(cap_context.eventloop)
> cap_context.eventloop.run_until_complete(asyncio.gather(*pending))
> ```

The second line in this code snippet iterates through the entire capture. This GitHub issue is about `pyshark` raising an exception when a capture is closed _before_ it is exhausted.

**Comment by wlockwood** [comment]
> This GitHub issue is about pyshark raising an exception when a capture is closed before it is exhausted.
I was getting this error even when exhausting the capture, hence ending up on this issue.

**Comment by KimiNewt** [comment]
Managed to recreate this on Windows. I made a branch & PR. Can you confirm that this fixes your issue? 
`fix-windows-eventloop-issues` - #566

**Comment by KimiNewt** [comment]
This seems like a separate issue. I'll upload a new version to pypi shortly and you can open an issue regarding the setup problem.
